lte/cloud/go/services/subscriberdb/storage/digest_lookup.go

```diff
@@ -0,0 +1,207 @@
+/*
+ Copyright 2020 The Magma Authors.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree.
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package storage
+
+import (
+	"database/sql"
+
+	"magma/orc8r/cloud/go/clock"
+	"magma/orc8r/cloud/go/sqorc"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/pkg/errors"
+)
+
+type digestLookup struct {
+	db      *sql.DB
+	builder sqorc.StatementBuilder
+}
+
+type DigestInfo struct {
+	networkID       string
+	digest          string
+	lastUpdatedTime int64
+}
+
+type DigestLookup interface {
+	// Initialize the backing store.
+	Initialize() error
+
+	// GetDigests returns a list of digests of that satisfy the filtering criteria
+	// specified by the arguments.
+	//
+	// If networkIDs is empty, return all up-to-date digests. If earliestUpdateTime
+	// is negative, return all digests of the selected networks regardless of last
+	// update time.
+	GetDigests(networkIDs []string, earliestUpdateTime int64) ([]DigestInfo, error)
+
+	// SetDigest creates/updates the subscribers digest for a particular network.
+	SetDigest(networkID string, digest string) error
+
+	// DeleteDigests removes digest rows by network IDs.
+	DeleteDigests(networkIDs []string) error
+}
+
+const (
+	digestLookupTableName = "subscriberdb_network_id_to_digest"
+
+	digestLookupNidCol             = "network_id"
+	digestLookupDigestCol          = "digest"
+	digestLookupLastUpdatedTimeCol = "last_updated_at"
+)
+
+func NewDigestLookup(db *sql.DB, builder sqorc.StatementBuilder) DigestLookup {
+	return &digestLookup{db: db, builder: builder}
+}
+
+func (l *digestLookup) Initialize() error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.CreateTable(digestLookupTableName).
+			IfNotExists().
+			Column(digestLookupNidCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			Column(digestLookupDigestCol).Type(sqorc.ColumnTypeText).NotNull().EndColumn().
+			Column(digestLookupLastUpdatedTimeCol).Type(sqorc.ColumnTypeBigInt).NotNull().EndColumn().
+			PrimaryKey(digestLookupNidCol).
+			RunWith(tx).
+			Exec()
+		return nil, errors.Wrap(err, "initialize Digest lookup table")
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+func (l *digestLookup) GetDigests(networkIDs []string, earliestUpdateTime int64) ([]DigestInfo, error) {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		filters := squirrel.And{}
+		if len(networkIDs) > 0 {
+			filters = append(filters, squirrel.Eq{digestLookupNidCol: networkIDs})
+		}
+		if earliestUpdateTime >= 0 {
```

Comment:
The conditional is `squirrel.Lt` which specifies that `digestLookupLastUpdatedTimeCol < earliestUpdateTime` (since it's filtering for outdated digests, not up-to-date ones). In the case where `earliestUpdateTime <= 0`, this conditional would be `false` for all digests. @hcgatewood